perlapi: Combine all av_len synonyms into one entry

khwilliamson opened the pull request.

khwilliamson committed:
autodoc.pl: White-space only

Wrap lines to fit in an 80 column terminal

khwilliamson committed:
autodoc: improve internal uniformity of entries

Originally, each entry was for a single function or macro, etc.  Then
the ability to group similar items in a single entry was shoe-horned in,
with the extra ones in a group being a subsidiary data structure with
most of the same elements.  But it is easier to code new things based on
a uniform structure with all items the same.  That's what this commit
does.

The first item in a group is added to the array of all the items as the
0th element.

This had already been done in one subroutine because it made for easier
handling; this commit moves that to much earlier in the process.

tonycoz commented:
> Only query is will it look a bit odd now to document the return type for just `AvFILL`?

The others have entries in `embed.fnc` so `autodoc.pl` picks up the return and parameter types from there.

khwilliamson committed:
autodoc: Extract code into a function

This is in preparation for it being called from a second place.

khwilliamson committed:
autodoc: Extract code into function

This is in preparation for it to be called from more places

khwilliamson committed:
autodoc: Add missing /x on pattern

The issue this was supposed to be protecting against had never come up

khwilliamson committed:
autodoc: Clarifications

This changes some comments and white space, and reverses the logic of
some conditionals to make clearer what is being done, especially with
future commits

khwilliamson committed:
autodoc: Add apidoc_defn directive

perlapi and perlintern document a variety of constructs.  Function
prototypes  are declared in embed.fnc and used to generate header files
that the compiler sees, for example, to declare function signatures
without having to worry that a function call will be encountered during
compilation prior to its definition.  The compiler likely notes any
discrepancies.

Macros don't have that kind of checking, beyond having the correct
number of parameters.  We have added some further macros that can be
applied to individual parameters in the main macro's definition to make
sure the parameter is a pointer when it should be, and vice versa.  But
lots of possible issues are uncaught.

We also can specify our expected types of each parameter in embed.fnc
like we do for functions, and additionally in apidoc lines.  But, those
can be wrong without it being apparent.

To lessen the chances of being wrong, it would be advantageous to place
the declaration of the macro's signature adjacent to its actual
definition in the code.  Discrepancies are easily noted.

But this conflicts with combining many variants of a function into the
same api entry.  Some variants will be functions, and their definitions
will likely be in .c files, while others will be macros that merely wrap
the function call.  These will necessarily be in .h files.

This commit is a solution.  The new construct 'apidoc_defn' is created.
This is to be placed adjacent to the macro's definition, documenting
there how we expect it to be called.  But it only contains the macro's
signature.  Somewhere else in the source will be a regular apidoc (or
apidoc_item) line without any signature information attached to it.
That line will put the information from the apidoc_defn no matter where
in the source it is located.

Prior to this commit, that was already true for macro signatures in
embed.fnc.  The apidoc line pulls the signature information from what
embed.fnc generates.  This essentially just supplements that information
from lines in the source corpus.

We could gradually move macro signatures from embed.fnc to be adjacent
to the actual macro defines.

khwilliamson committed:
autodoc: Process .h files before .c ones

The previous commit adds code that takes macro signature definitions in
the source and allows them to be used from elsewhere.  When a use comes
before the definition it has to be saved for later processing.  When a
definition comes before the use, it can be processed immediately.  Most
definitions will be in header files.  It is a simple matter to arrange
these to be processed before any .c files, thus minimizing the number of
instances of having to save for later.

khwilliamson committed:
perlapi: Combine all av_len synonyms into one entry

khwilliamson commented:
I realized that it would be better if the issue about the signatures didn't come up at all.  So, I changed autodoc to allow a signature declaration line next to the macro definition which will be pulled in by the actual autodoc line.

tonycoz commented on autodoc.pl:
```diff
@@ -1797,12 +1799,21 @@ sub add_defn  {
-    open F, '<', $file or die "Cannot open $file for docs: $!\n";
-    autodoc(\*F,$file);
-    close F or die "Error closing $file: $!\n";
+    if (0 && $file =~ /\.h/) {
```
Condition always false, presumably left over from debugging.

khwilliamson replied:
yes; now fixed

tonycoz reviewed (approved):
This looks fine.

I wondered if there should be an update to documentation for the autodoc pod syntax, but I didn't find anything to update.

khwilliamson commented:
The new syntax is documented in `embed.fnc`
